### DISCUSSION TITLE
Diesel generated a new database file instead of pointing to the intended one

### DISCUSSION THREAD

**Discussion by rsteaburdea** [diesel-rs/diesel#3604] [Q&A]
### Issue description
The issue that I am facing is that Diesel generated a new file at this path: `/Users/My Username/Library/Application Support/com.tauri.dev/pownshop.sqlite3` instead of pointing to the one that I created.
Another problem that I am facing is with the migration scripts that doesn't update the `pownshop.sqlite3` database that was created. (after running `diesel migration redo` and `diesel migration run`)

Here is a little [demo](https://drive.google.com/file/d/1gKdDk8w7qfEEHsYln6M4Ep8HkuE-QuPN/view) with the issue.
I took inspiration from [this app](https://github.com/TristanPouliquen/selection-poste-enm).

The `.env` file looks like this (it doesn't work with absolute path either):
```
TAURI_DEV_WATCHER_IGNORE_FILE=.taurignore
DATABASE_URL=../../pownshop.sqlite3
```

The `.taurignore` file looks like this:
```
db/pownshop.sqlite3
db/pownshop.sqlite3-journal
```

The code looks like this:
```
pub fn establish_connection(path: &str) -> SqliteConnection {
    SqliteConnection::establish(path).unwrap_or_else(|_| panic!("Error connecting to database"))
}

#[derive(Identifiable, Queryable, Serialize, Deserialize, AsChangeset)]
#[serde(rename_all = "camelCase")]
#[diesel(table_name = pwn_contracts)]
pub struct Contract {
    pub id: i32,
    pub name: String,
}

pub fn contract_list(db_path: String) -> Vec<Contract> {
    let mut connection = establish_connection(&db_path);
    pwn_contracts::dsl::pwn_contracts
        .select(pwn_contracts::all_columns)
        .load::<Contract>(&mut connection)
        .expect("Loading contracts failed")
}
```

In `main.rs`
```
pub const MIGRATIONS: EmbeddedMigrations = embed_migrations!("./migrations");
pub const DATABASE_NAME: &str = "pawnshop.sqlite3";

fn get_db_path(app_handle: tauri::AppHandle) -> String {
    let mut app_dir = app_handle.path_resolver().app_data_dir().unwrap();
    app_dir.push(DATABASE_NAME);
    app_dir.display().to_string()
}

#[tauri::command]
fn get_contracts(app_handle: tauri::AppHandle) -> Vec<Contract> {
    let db_path = get_db_path(app_handle);
    contract_list(db_path)
}

.setup(|app| {
    let mut dir_path = app.path_resolver().app_data_dir().unwrap();
    let _ = std::fs::create_dir_all(&dir_path);
    dir_path.push(DATABASE_NAME);
    let path = dir_path.display().to_string();
    let connection = &mut establish_connection(&path);
    connection
        .run_pending_migrations(MIGRATIONS)
        .expect("Error running pending migrations");
    Ok(())
})
```

The `Cargo.toml` dependencies are this:
```
[dependencies]
serde_json = "1.0"
serde = { version = "1.0", features = ["derive"] }
tauri = { version = "1.2.4", features = ["shell-open"] }
diesel = { version = "2.0.0", features = ["sqlite", "returning_clauses_for_sqlite_3_35"] }
libsqlite3-sys = { version = "0.25.2", features = ["bundled"] }
diesel_migrations = {version = "2.0.0", features = ["sqlite"] }
remove_dir_all = "0.8.2"
```

### Questions
1. How to I make the app point to the database that I am creating manually instead of the generated one?
2. How do I make the migration scripts work properly?

### Platform and versions
```
Environment
  › OS: Mac OS 13.3.1 X64
  › Node.js: 18.16.0
  › npm: 9.6.4
  › pnpm: 8.2.0
  › yarn: 1.22.19
  › rustup: 1.25.2
  › rustc: 1.68.2
  › cargo: 1.68.2
  › Rust toolchain: stable-aarch64-apple-darwin

Packages
  › @tauri-apps/cli [NPM]: 1.2.3
  › @tauri-apps/api [NPM]: 1.2.0
  › tauri [RUST]: 1.2.4,
  › tauri-build [RUST]: 1.2.1,
  › tao [RUST]: 0.15.8,
  › wry [RUST]: 0.23.4,

App
  › build-type: bundle
  › CSP: unset
  › distDir: ../dist/pownshop-tauri
  › devPath: http://localhost:1420/
  › framework: Angular
  › bundler: Webpack

App directory structure
  ├─ dist
  ├─ .angular
  ├─ node_modules
  ├─ src-tauri
  ├─ .vscode
  └─ src
  ```
  
  ## Issue Updates
  **20.04.2023 - 08:57 (GMT + 3)**
  Added a video with a demo and updated the code section for a more relevant description

**Comment by weiznich** [comment]
Given the provided code it's unclear how `establish_connection` is called. This function will open a connection to a database at the provided `path`. I would check the provided path and if it does not match the expected value, check how this function is called.

**Reply by weiznich** [reply]
> Another problem that I am facing is with the migration scripts that doesn't update the pownshop.sqlite3 database that was created. (after running diesel migration redo and diesel migration run)

Migrations that are applied will not be applied again to the same database, at least not without reverting them first.

**Reply by rsteaburdea** [reply]
I've added here a link with a little [demo](https://drive.google.com/file/d/1gKdDk8w7qfEEHsYln6M4Ep8HkuE-QuPN/view?usp=sharing), maybe it will help.

Can you have a look please

**Reply by weiznich** [reply]
Well, given the provided code: Diesel will create the database there to whatever `path` in your setup handle evaluates. That's strictly speaking not in diesels control and as I don't know tauri in details I cannot really help you here with details. Maybe try checking their support forums for help how to get the path you would expect there?

**Reply by rsteaburdea** [reply]
I already did that and I talked with someone from support on Tauri's discord server and he said : 

>yeah looks like the path is messed up. in your main function you're setting `$HOME/Library/Application Support/<bundle-identifier>/` as the base dir, and append `DATABASE_NAME` to that, meaning the database could also be in: 
`$HOME/Library/Application Support/<bundle-identifier>/Users/rsteaburdea/Workspace/pownshop/pownshop.sqlite3`
 or something like that
>
> Anyway, tauri can't read .env files itself and while there are crates for it, using DATABASE_URL like this would not work in a production app, at least if you ship it to users. If you _do_ want to ship your app to users, i'd suggest to just stick with the `Application Support` path you have in your code.
> 
> If you want to read the DATABASE_URL env var in development, you could look into this crate: <https://docs.rs/envy/latest/envy/>
> 
> The code itself (get_db_path) will work fine, but whatever diesel reads the .env file for will fail, no idea what that will be though

For me it felt that it was a diesel issue because of the database file path and I decided to post it here.

1. How do I revert the migrations? Shouldn't this command from Diesel CLI fix it: `diesel migration redo`?
2. Why is this database generated at this path: `/Users/My Username/Library/Application Support/com.tauri.dev/pownshop.sqlite3`?
3. How do I make it point to the `.sqlite3` file that was created by me?

**Reply by weiznich** [reply]
I'm feeling this leading to nowhere, as you don't provide the information that are required to help you here. Generally speaking: I usually expect that people do provide an reproducible (ideally minimal) example of their problem, especially if they ask others to solve their problems in their free time. As this question does not contain such an example let me write down a few general points:

* Diesel itself does create the database exactly where you instruct in the database to be created. So if you pass in `/whatever/path/foo.db` it will create the db there. There is no additional code going on there and it will not read any `.env` file (that needs to be done explicitly by your application).
* Diesel-CLI uses the `--database-url` parameter or if that is not set the `DATABASE_URL` environment variable get the path. The later will use `.env` files.
* `diesel migration redo` will revert and rerun the last migration. Generally speaking, `diesel-cli` will not do anything automatically, so if you change a specific migration you explicitly need to rerun it via redo. If that's not the latest migration you need to use the `-n N` flag to instruct it to run the last `N` number of migrations.

For additional support: Please provide a minimal reproducible example of your problem. Ideally without dependencies on any third party crate.

**Reply by rsteaburdea** [reply]
Not trying to be rude, just wanted to point that I've attached to this post a short video with me reproducing the bug and a repo from where I took inspiration. Also I don't think it is productive to copy paste the entire code. Also I understand the meaning of using free time to fix ofhers problema and is very much appreciated.

Could you be more specific please regarding the details you need in order to understand the problem?

**Reply by weiznich** [reply]
Neither the video nor the example repo you've based your code on is a reproducible example. That would be something that I can clone and run a limited number of steps in to see your issue. As written above: It should be stripped down as much as possible form your real application. I know that it is work to produce something like that, but thats the least thing you could do for other people that try to help you.

**Reply by rsteaburdea** [reply]
Ok, I will strip it down as much as posible asap and attach a github repo. Thank you
